### PR TITLE
Standardize For Loops

### LINKED ISSUE #2: Update range based loops to use auto where appropriate

### LINKED ISSUE #8: Standardize for loops to use ++/-- i (unary operator before the variable)

### PR DESCRIPTION
**Pull request by Brett208**
- Update to newest version of OP2Utility submodule
 - Increase use of range based for loops and auto keyword in for loops
 - Make ConsoleArgumentParser::ArgumentMatch a const function
 - Remove unused function prototype ConsoleCreate::ComparePathFilenames

Closes #2 
Closes #8

### CODE REVIEW

**Comment by Brett208** [comment]
I'm not exactly sure what happens here. OP2Archive is updating to branch Standardize For Loops in OP2Utility. So when the branch is merged and deleted in OP2Utility, I'm not sure where that leaves this pull request. I'm guessing we need to go in and re-update this one to OP2Utility Master at some point?

-Brett

File: ConsoleRemove.cpp [hunk 1]
```diff
@@ -87,17 +87,16 @@ void ConsoleRemove::CheckFilesAvailableToRemove(ArchiveFile& archive, const vect
...
 void ConsoleRemove::ExtractFilesFromOriginalArchive(ArchiveFile& archive, const vector<string> internalFilenames)
 {
-	for (std::size_t i = 0; i < internalFilenames.size(); ++i)
+	for (const auto& internalFilename : internalFilenames)
 	{
-		string filename(internalFilenames[i]);
-		int index = archive.GetInternalFileIndex(filename.c_str());
-		string pathToExtractTo = XFile::AppendSubDirectory(filename, tempDirectory);
+		int index = archive.GetInternalFileIndex(internalFilename.c_str());
+		string pathToExtractTo = XFile::AppendSubDirectory(internalFilename, tempDirectory);
```
Comment (DanRStevens) [inline ConsoleRemove.cpp]: This is correct, though stood out because the API is still using `const char*` rather than `const std::string&`.

Additionally, we may want to consider if the interface provided by `ArchiveUnpacker` is really meeting our needs. It seems odd for the library user to need to manually lookup a filename and translate it to an index, and then extract the file using the index. Certainly there are cases where this flexibility is useful, but we also want the common case to be easy, while also not requiring library users to be familiar with a ton of functions. We may be able to make use of functions that compose better. In particular, we may soon be able to make use of the method to copy one stream to another as a flexible way to extract files.

Comment (Brett208) [reply ConsoleRemove.cpp]: Actually, the API is using std::string now. C++ does not provide any warnings or errors when you convert from string to char* and then back to string. These calls had to convert to char* back when it was required by the Archives code. Now that we have updated to std::string, it continued to work and I missed removing the c_str() calls. I was generally aware it was happening during the loop cleanup and was going to Grep c_str to try to remove all the old calls but you beat me to pointing it out. :)

Agreed it would be a nice feature to be able to extract a file based on its internal filename in the volume. Then usage could be something like:

```if (archive.ContainsFile(filename) {
    archive.extractFile(filename, pathForExtraction);
}
```

While this is an easy fix, lets open an issue on it since I don't have time to fix right this moment. Also, I just realized ArchiveUnpacker::ContainsFile still uses a char* and needs to be updated to a string.

**Comment by Brett208** [comment]
I'm not really sure how to reference a specific commit hash. Would that mean when fetching code, OP2Archive would be smart enough to grab the master branch if the commit was moved from the temp branch to the master?

Other thing I wonder is if rebase and merge changes the commit hash and maybe needs to be avoided in these situations?

For now I'll merge OP2Utility into the master first, then update this branch to the master of OP2Utility and then merge it. Lets try targeting the hash on the next branch that it comes up on though.

Thanks for the review,
-Brett

File: ConsoleRemove.cpp (hunk 1 above)
Comment (DanRStevens) [reply ConsoleRemove.cpp]: Yes indeed, ArchiveUnpacker has a number of methods still using `const char *`.

**Comment by DanRStevens** [comment]
How are you specifying the submodule? I assume initial setup needs a repo, and from then on a specific commit must be referenced on each update. That commit might be through a symbolic name, such as the branch name, or HEAD, or through the SHA. I'm saying it should be specified with an SHA rather than a branch name. Ideally, the SHA should be on master.

If you did a rebase and merge, it may rewrite history, which may change the commit SHA. I believe the SHA may be preserved if nothing else has been pushed to master since either the branch point or the most recent rebase. If you do a squash and merge, the SHA should be preserved if nothing else has been pushed to master since either the branch point or the most recent rebase, and the branch contains only a single commit. If you do a merge commit, that SHA should be preserved on master, but will be down one of the forked paths, unless a fast-forward merge was done, which I don't think is default behaviour on GitHub. For a merge commit, I think it would be more proper to wait until the merge was actually done, and then use the SHA of the merge commit itself, rather than the commit before which is being merged into master. If you use the branch SHA, you may miss changes that were incorporated into master after the branch was started or last rebased.

In short, you should be merging into master before updating a submodule reference to point at the new code.

File: ConsoleRemove.cpp (hunk 1 above)
Comment (DanRStevens) [reply ConsoleRemove.cpp]: Hmm, my bad. I was looking at out of date code. I had fetched changes, but not merged them into my local master branch. There is still one method that takes `const char*` though:
```
virtual std::unique_ptr<SeekableStreamReader> OpenSeekableStreamReader(const char *internalFileName) = 0;
```

**Comment by Brett208** [comment]
Makes sense. Thanks.
